js/plugins/mcp/src/client/client.ts

```diff
@@ -0,0 +1,291 @@
+import { Genkit, GenkitError, ToolAction } from 'genkit';
+import { logger } from 'genkit/logging';
+import { transportFrom } from '../util';
+import { fetchDynamicResourceTools } from '../util/resources';
+import { fetchDynamicTools } from '../util/tools';
+export type { SSEClientTransportOptions, StdioServerParameters, Transport };
+
+interface McpServerRef {
+  client: Client;
+  transport: Transport;
+  error?: string;
+}
+
+export interface McpClientControls {
+  // Optional server name
+  name?: string;
+  // when true, the server will be stopped and its registered components will not appear in lists/plugins/etc
+  disabled?: boolean;
+}
+
+export type McpStdioServerConfig = StdioServerParameters;
+
+export type McpSSEServerConfig = { url: string } & SSEClientTransportOptions;
+
+export type McpTransportServerConfig = {
+  transport: Transport;
+  client: Client;
+  command?: never;
+  url?: never;
+};
+
+/**
+ * Configuration for an individual MCP server. The interface should be familiar
+ * and compatible with existing tool configurations e.g. Cursor or Claude
+ * Desktop.
+ *
+ * In addition to stdio servers, remote servers are supported via URL and
+ * custom/arbitary transports are supported as well.
+ */
+export type McpServerConfig =
+  | McpStdioServerConfig
+  | McpSSEServerConfig
+  | McpTransportServerConfig;
+
+/**
+ * Configuration options for an individual `GenkitMcpClient` instance.
+ * This defines how the client connects to a single MCP server and how it behaves.
+ */
+export interface McpClientOptions extends McpClientControls {
+  /**
+   * An optional version number for this client. This is primarily for logging
+   * and identification purposes. Defaults to '1.0.0'.
+   */
+  version?: string;
+  /**
+   * The configuration for the MCP server to which this client will connect.
+   * This includes details like the server's command, URL, or a pre-existing transport.
+   * See `McpServerConfig` for more details.
+   */
+  server: McpServerConfig;
+  /**
+   * If true, tool responses from the MCP server will be returned in their raw
+   * MCP format. Otherwise (default), they are processed and potentially
+   * simplified for better compatibility with Genkit's typical data structures.
+   */
+  rawToolResponses?: boolean;
+}
+
+/**
+ * Represents a client connection to a single MCP (Model Context Protocol) server.
+ * It handles the lifecycle of the connection (connect, disconnect, disable, re-enable, reconnect)
+ * and provides methods to fetch tools from the connected server.
+ *
+ * An instance of `GenkitMcpClient` is typically managed by a `GenkitMcpClientManager`
+ * when dealing with multiple MCP server connections.
+ */
+export class GenkitMcpClient {
+  private name: string;
+  private version: string;
+  private serverConfig: McpServerConfig;
+  private rawToolResponses?: boolean;
+  private disabled: boolean;
+
+  private _server?: McpServerRef;
```

Comment:
Let's make this public so that there's an escape hatch for manually doing things with the underlying MCP SDK.